src/SOS/SOS.UnitTests/Debuggees/DotnetDumpCommands/Program.cs

```diff
@@ -127,6 +127,11 @@ private static IEnumerable<object> CreateObjectsInDifferentGenerations()
             // This object should go into LOH
             yield return new DumpSampleClass[50000];
 
+#if NET5_0_OR_GREATER
```

Comment:
I'm not sure if this is the correct way to only test this on .NET 5+.

Reply:
[SOS.DumpGen tests](https://dev.azure.com/dnceng/public/_build/results?buildId=1127370&view=ms.vss-test-web.build-test-results-tab&runId=34258674&resultId=100167&paneView=debug) fail for .NET 5 and .NET 6 on Windows-x86.

```
System.Exception : Debugger output did not match the expression: ^[A-Fa-f0-9]+(`[A-Fa-f0-9]+)?\s+1\s+10024\s+System.Byte\[\]
```
The pattern itself is correct (verifiyed with a regex tool and the sample code from https://github.com/dotnet/diagnostics/pull/2260#discussion_r628114030).

Looking at the assemblies from the artifacts, I see that the conditional compilation for > .NET 5 works.

.NET 5 and .NET 6
![grafik](https://user-images.githubusercontent.com/5755834/117444237-78d2a700-af39-11eb-936d-b63cf6d820c1.png)

Previous versions:
![grafik](https://user-images.githubusercontent.com/5755834/117444272-88ea8680-af39-11eb-964c-c1325546ed2e.png)

From the test-logs:
```
...
00:00.187: Writing full to D:\workspace\_work\1\s\artifacts\tmp\Release\dumps\ProjectK\6.0.0-preview.5.21254.12\net6.0\SOS.DotnetDumpCommands.Heap.dmp
...
00:00.536: Loading Dump File [D:\workspace\_work\1\s\artifacts\tmp\Release\dumps\ProjectK\6.0.0-preview.5.21254.12\net6.0\SOS.DotnetDumpCommands.Heap.dmp
...

STDIN: 00:00.504: dumpgen poh
       00:00.504: > dumpgen poh
       00:00.504: Statistics:
       00:00.504:       MT    Count    TotalSize Class Name
       00:00.504: 03350e58        1           12 Free
       00:00.504: Total 1 objects
       00:00.504: 
       00:00.504: <END_COMMAND_OUTPUT>

...
```
so the POH-allocation doesn't show up (in CI). 

I haven't installed x86 (at the moment), so didn't verify it locally.

The same test on Windows-x64-Release:
```
...
STDIN: 00:00.562: dumpgen poh
00:00.562: > dumpgen poh
00:00.562: Statistics:
00:00.562:               MT    Count    TotalSize Class Name
00:00.562: 000001923afa71f0        1           24 Free
00:00.562: 00007ffc8e05d578        1        10024 System.Byte[]
00:00.562: 00007ffc8df7b070        3        17416 System.Object[]
00:00.562: Total 5 objects
00:00.562: 
00:00.562: <END_COMMAND_OUTPUT>
...
```
passes.

Has the GC / POH different behaviour on x86 and x64 here?

Reply:
I see that you already from the script conditional.

Reply:
@Maoni0 here's another question, please.

The test-code does `GC.AllocateUninitializedArray<byte>(10000, pinned: true);` and then a dump is taken.
In the x64-test everything is fine, this objects is listed in `dumpgen poh` (new command by this PR, relying on ClrMD).
But the x86-test fails as this pinned object doesn't show up by `dumpgen poh`.

Are there differences between x86/x64 for the POH? (I'm only aware of the managed side of AllocateUninitializedArray, not the native one).

If there's no difference the cause will be somewhere else...

Reply:
it should work on x86. I would suggest to use sos to dump what's on the POH and verify that the objects are on POH (they should). I just verified this and sos correctly shows objs on POH for both x86 and x64.

Reply:
Thanks for the verification!
I'll need to dig deeper what's going on here (later when I'm in the office).

Reply:
I tried a simple example (x86 -- note: x64 works fine):
```c#
using System;

var buffer = GC.AllocateUninitializedArray<byte>(10_000, pinned: true);

Console.WriteLine("Dump now");      // dump taken here
Console.ReadKey();

Console.WriteLine(buffer.Length);
```
Analyzing the heap with SOS shows:
```
> dg gen2
Statistics:
      MT    Count    TotalSize Class Name
0074c830        1           12 Free
Total 1 objects

> dg poh
Statistics:
      MT    Count    TotalSize Class Name
0074c830        1           12 Free
Total 1 objects
```
So nowhere is the pinned byte[].

```
> dumpheap -stat
Statistics:
      MT    Count    TotalSize Class Name
07d0d0d4        1           12 System.Threading.Tasks.Task+<>c
...
00d0b9d0        7         2020 System.Int32[]
00d045e8       10         9364 System.Object[]
07d000d0        5        13695 System.Byte[]
00d0ec98      142        37048 System.String
Total 333 objects

> dumpheap -mt 07d000d0
 Address       MT     Size
048ca30c 07d000d0       29
048ca664 07d000d0       46
048cad24 07d000d0       12
048cbe90 07d000d0     3596
068c3220 07d000d0    10012

Statistics:
      MT    Count    TotalSize Class Name
07d000d0        5        13695 System.Byte[]
Total 5 objects
```
Here it is at `068c3220`, so
```
> gcwhere 068c3220
Address    Gen   Heap   segment    begin      allocated   size
068C3220   4      0     068C0000   068C1000   068C5940    0x271c(10012)
```
shows that this byte[] is in the POH -- correct, and as expected.

Let's start a debug-session for `dotnet-dump analyze` and execute `dg gen2` and `dg poh`. Looking at the segment's start and end at
https://github.com/dotnet/diagnostics/blob/c13a550fd8135e031678015743557ea2a543403f/src/Microsoft.Diagnostics.ExtensionCommands/ClrMDHelper.cs#L715
I see:
```
Gen 2
start = 0x00000000048c1000
end   = 0x00000000048c100c

LOH
start = 0x00000000058c1000
end   = 0x00000000058c1010

POH
start = 0x00000000068c1000
end   = 0x00000000068c5940
```
These values match the ones from SOS:
```
> eeheap -gc
Number of GC Heaps: 1
generation 0 starts at 0x048C1018
generation 1 starts at 0x048C100C
generation 2 starts at 0x048C1000
ephemeral segment allocation context: none
 segment     begin  allocated  committed  allocated size  committed size
048C0000  048C1000  048CD9D4  048D2000  0xc9d4(51668)  0x11000(69632)
Large object heap starts at 0x058C1000
 segment     begin  allocated  committed  allocated size  committed size
058C0000  058C1000  058C1010  058C2000  0x10(16)  0x1000(4096)
Pinned object heap starts at 0x068C1000
068C0000  068C1000  068C5940  068D2000  0x4940(18752)  0x11000(69632)
Total Allocated Size:              Size: 0x11324 (70436) bytes.
Total Committed Size:              Size: 0x12000 (73728) bytes.
------------------------------
GC Allocated Heap Size:            Size: 0x11324 (70436) bytes.
GC Committed Heap Size:            Size: 0x12000 (73728) bytes.
```

Dumping the heap just for the POH-range:
```
> dumpheap 0x00000000068c1000 0x00000000068c5940
 Address       MT     Size
068c1000 0074c830       12 Free
068c1010 00d045e8     4092
068c2010 00d045e8      524
068c2220 00d045e8     4092
068c3220 07d000d0    10012

Statistics:
      MT    Count    TotalSize Class Name
0074c830        1           12      Free
00d045e8        3         8708 System.Object[]
07d000d0        1        10012 System.Byte[]
Total 5 objects
```
The byte[] array shows up, as expected.

Stepping with the debugger through `dg poh` and focusing on https://github.com/dotnet/diagnostics/blob/c13a550fd8135e031678015743557ea2a543403f/src/Microsoft.Diagnostics.ExtensionCommands/ClrMDHelper.cs#L703-L711

I can see this for the first object:
![grafik](https://user-images.githubusercontent.com/5755834/117537000-91f15b80-affe-11eb-8001-3cbb1c45b4ca.png)
In the next iteration (should be the object[]):
![grafik](https://user-images.githubusercontent.com/5755834/117537050-ea285d80-affe-11eb-93d8-74326409e089.png)
So `currentObject.Type = null` (which shouldn't be?!).
Then in the next step
![grafik](https://user-images.githubusercontent.com/5755834/117537086-1a6ffc00-afff-11eb-93d2-39eef0dccc42.png)
`currentObjectAddress = nullptr` and the loop exits.

Doing the same debugging-sequence on x64 with x64-dump, then the objects are enumerated correctly.

So it looks like there's a bug in this loop for x86.
At the moment I'm not familiar with the code enough to spot and fix it.
(Will look into this later...)